include/deal.II/base/mpi_noncontiguous_vector.h

```diff
@@ -0,0 +1,256 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2020 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_mpi_noncontiguous_vector_h
+#define dealii_mpi_noncontiguous_vector_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/mpi.h>
+#include <deal.II/base/mpi.templates.h>
+#include <deal.II/base/mpi_compute_index_owner_internal.h>
+#include <deal.II/base/mpi_tags.h>
+
+#include <deal.II/lac/communication_pattern_base.h>
+#include <deal.II/lac/vector_space_vector.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Utilities
+{
+  namespace MPI
+  {
+    /**
+     * A flexible Partitioner class, which does not make an restrictions
+     * regarding the order of the underlying index sets.
+     *
+     * @author Peter Munch, 2020
+     */
+    template <typename Number = double>
+    class NoncontiguousPartitioner
+      : public LinearAlgebra::CommunicationPatternBase
+    {
+    public:
+      /**
+       * Constructor. Setup point-to-point communication pattern based on the
+       * IndexSets @p indexset_has and @p indexset_want for the MPI communicator
+       * @p communicator.
+       */
+      NoncontiguousPartitioner(const IndexSet &indexset_has,
+                               const IndexSet &indexset_want,
+                               const MPI_Comm &communicator);
+
+      /**
+       * Constructor. Same as above but vectors of indices @p indices_has
+       * and @p indices_want. This allows that the indices do not have to be
+       * sorted and the values are read and written automatically at the right
+       * position of the vector during update_values(), update_values_start(),
+       * and update_values_finish().
+       */
+      NoncontiguousPartitioner(
+        const std::vector<types::global_dof_index> &indices_has,
+        const std::vector<types::global_dof_index> &indices_want,
+        const MPI_Comm &                            communicator);
+
+      /**
+       * Fill the vector @p dst according to the precomputed communication
+       * pattern with values from @src.
+       *
+       * @pre The vectors only have to provide a method data(), which allows
+       *   to access their raw data.
```

Comment:
I do not think this is what is used in the function - I see `operator[]` to be used. Maybe change it? Another question is whether we should change the necessary interface to `begin()` and `end()` to also work with the local range of `LA::d::Vector` (not sure if it would make sense directly).

Reply:
You are right. `data()` is needed by the buffer. I like the idea with `begin()` and have added this.